### PR TITLE
ref #870 - improves memory management

### PR DESCRIPTION
**Pull request by Evelf**
Multiple imbricated loops are needs to generate the compressed files and manifest, but the way it's currently done is wasteful in memory when we have multiple contexts.

With a few meticulous changes, it can be drastically improved, in my benchmarks it goes from \~2942Mb to \~174Mb in memory consumption with no impact on the execution time (\~ 3'15"). See the attached mprofile files.

[mprofile.zip](https://github.com/django-compressor/django-compressor/files/1228841/mprofile.zip)

### CODE REVIEW

File: compressor/management/commands/compress.py [hunk 1]
```diff
@@ -287,6 +285,9 @@ def handle_extensions(self, extensions=('html',)):
         return set(ext_list)
 
     def handle(self, **options):
+        self.handle_inner(**options)
+
+    def handle_inner(self, **options):
```
Comment (karyon) [inline compressor/management/commands/compress.py]: in how far does this help?

File: compressor/offline/jinja2.py [hunk 2]
```diff
@@ -84,7 +84,7 @@ def get_init_context(self, offline_context):
 
         # Allow offline context to override the globals.
         context = self.env.globals.copy()
-        context.update(offline_context)
+        context.update(flatten_context(offline_context))
```
Comment (karyon) [inline compressor/offline/jinja2.py]: could you explain this change?

File: compressor/tests/test_offline.py [hunk 3]
```diff
@@ -560,21 +562,10 @@ class OfflineCompressTestCaseErrors(OfflineTestCaseMixin, TestCase):
...
-            # Because we use env.parse in Jinja2Parser, the engine does not
-            # actually load the 'extends' and 'includes' templates, and so
-            # it is unable to detect that they are missing. So all the
-            # 'compress' nodes are processed correctly.
-            self.assertEqual(4, count)
-            self.assertEqual(engine, 'jinja2')
-            self.assertIn(self._render_link('7ff52cb38987'), result)
-            self.assertIn(self._render_link('2db2b4d36380'), result)
-
+        self.assertEqual(2, count)
```
Comment (karyon) [inline compressor/tests/test_offline.py]: could you explain this?

File: compressor/management/commands/compress.py (hunk 1 above)
Comment (Evelf) [inline compressor/management/commands/compress.py]: We need ```handle_inner()``` return values for the tests, but Django's ```handle()``` can not return anything.

File: compressor/tests/test_offline.py (hunk 3 above)
Comment (Evelf) [inline compressor/tests/test_offline.py]: Seems that this PR "fixes" a bug in compressor management of jira's templates, there's only 2 non bugged templates and compressor should return only 2 results for both engines.

But I wanted to look a bit more closely at this, and forgot to do so before pushing the PR, thanks for pointing this again to me :)

File: compressor/offline/jinja2.py (hunk 2 above)
Comment (Evelf) [inline compressor/offline/jinja2.py]: It's a relic of the first attempt to improve memory consumption, it's needed to be able to use a generator for the context function.
It may or may not be part of this PR, as you prefer. In the end, that's not how performance where improved, but it's nice that you can give a generator to compressor without errors.

**Comment by karyon** [comment]
continued in #884, thanks :)

File: compressor/tests/test_offline.py (hunk 3 above)
Comment (karyon) [reply compressor/tests/test_offline.py]: this happened due to an introduced bug: the call to `compress` that got changed to `handle` required changing the parameter `engine` to `engines`, otherwise the parameter `engine` was not set, defaulted to `django`, and `jinja2` wasn't tested at all.
